app/models/user_password.rb

```diff
@@ -1,19 +1,80 @@
 # frozen_string_literal: true
 
 class UserPassword < ActiveRecord::Base
-  validates :user_id, presence: true
+  TARGET_PASSWORD_ALGORITHM =
+    "$pbkdf2-#{Rails.configuration.pbkdf2_algorithm}$i=#{Rails.configuration.pbkdf2_iterations},l=32$"
+  PASSWORD_SALT_LENGTH = 16
+  MAX_PASSWORD_LENGTH = 200
 
-  validates :user_id,
-            uniqueness: {
-              scope: :password_expired_at,
-            },
-            if: -> { password_expired_at.nil? }
+  attr_reader :password # required to build an instance of this model with password attribute without backing column
+  attr_accessor :salt # TODO: Deprecate once we drop User.salt, this is only for passing through the randomized salt from User
+  attr_accessor :should_expire_previous_passwords
+
+  belongs_to :user, required: true # different from user_id, we need to allow for blank user_id at the point of validation since a newly built user needs to be saved before user_id is generated
 
   validates :password_hash, presence: true, length: { is: 64 }, uniqueness: { scope: :user_id }
   validates :password_salt, presence: true, length: { is: 32 }
   validates :password_algorithm, presence: true, length: { maximum: 64 }
 
-  belongs_to :user
+  validate :password_validator
+  # validate :ensure_single_unexpired_password_for_user
+  before_save :expire_all_previous_passwords_for_user
```

Comment:
An expired password means

"This password still works, but you must change it on next login".

We shouldn't do that for all previous passwords. It's only intended for use in unusual situations like data leaks. 

So I think we should probably be destroying previous user-passwords instead? Or maybe even just updating the existing record in-place?

Would be interesting to see what we do for UserEmail here, which has a similar concept of a 'primary email', which can be changed using `user.email=`.